src/main/java/mate/academy/springboot/datajpa/controller/ProductController.java

```diff
@@ -0,0 +1,90 @@
+import mate.academy.springboot.datajpa.service.ProductService;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.web.bind.annotation.DeleteMapping;
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.PutMapping;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RequestParam;
+import org.springframework.web.bind.annotation.RestController;
+
+@RestController
+@RequestMapping("/products")
+public class ProductController {
+    private final CategoryService categoryService;
+    private final ProductService productService;
+    private final ResponseDtoMapper<ProductResponseDto, Product> productResponseDtoMapper;
+    private final RequestDtoMapper<ProductRequestDto, Product> productRequestDtoMapper;
+
+    @Autowired
+    public ProductController(
+            CategoryService categoryService,
+            ProductService productService,
+            ResponseDtoMapper<ProductResponseDto, Product> productResponseDtoMapper,
+            RequestDtoMapper<ProductRequestDto, Product> productRequestDtoMapper) {
+        this.categoryService = categoryService;
+        this.productService = productService;
+        this.productResponseDtoMapper = productResponseDtoMapper;
+        this.productRequestDtoMapper = productRequestDtoMapper;
+    }
+
+    @PostMapping
+    public ProductResponseDto create(@RequestBody ProductRequestDto productRequestDto) {
+        Product product = productRequestDtoMapper.toModel(productRequestDto);
+        Category category = categoryService.getByName(product.getCategory().getName());
+        category = category != null ? category : product.getCategory();
+        product.setCategory(categoryService.save(category));
+        return productResponseDtoMapper.toDto(productService.save(product));
+    }
+
+    @GetMapping
+    public ProductResponseDto getById(@RequestParam(name = "product_id") Long productId) {
+        return productResponseDtoMapper.toDto(productService.getById(productId));
+    }
+
+    @DeleteMapping
+    public void deleteById(@RequestParam(name = "product_id") Long productId) {
+        productService.delete(productService.getById(productId));
+    }
+
+    @PutMapping
+    public ProductResponseDto update(@RequestParam(name = "product_id") Long productId,
+                                     @RequestBody ProductRequestDto productRequestDto) {
+        Product product = productRequestDtoMapper.toModel(productRequestDto);
+        product.setId(productId);
+        Category category = categoryService.getByName(product.getCategory().getName());
+        category = category != null ? category : product.getCategory();
+        category.setName(product.getCategory().getName());
```

Comment:
you should do this in Product mapper, also, you should pass id of category, not its name

Reply:
RequestDto mustn't have id of entity